File: chapter-2-flink-streaming/src/main/scala/com/cnblogs/duma/streaming/scala/starter/WindowWordCount.scala
package com.cnblogs.duma.streaming.scala.starter

import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time

object WindowWordCount {
  def main(args: Array[String]) {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    val text = env.socketTextStream("192.168.29.132", 9000)

    val counts = text.flatMap { _.toLowerCase.split(" ") filter { _.nonEmpty } } //将输入数据按空格分割，过滤空元素
      .map { (_, 1) } //将word转成(word, 1)二元组
      .keyBy(0) //按照word分组
      .timeWindow(Time.seconds(5)) //5s的时间窗口
      .sum(1) //求和

    counts.print()

    env.execute("Window Stream WordCount")
  }
}

File: chapter-2-flink-streaming/src/main/java/com/cnblogs/duma/streaming/starter/WindowWordCount.java
package com.cnblogs.duma.streaming.starter;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

/**
 * @author duma http://www.duma.cnblogs.com
 */
public class WindowWordCount {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStream<Tuple2<String, Integer>> dataStream = env
                //创建 socket 数据源
                .socketTextStream("192.168.29.132", 9000)
                //扁平化(flat)操作，输入"a b c"，按照空格分割后，输出三个二元组 (a, 1) (b, 1) (c, 1)
                .flatMap(new Splitter())
                //按照元组的第一个属性分组
                .keyBy(0)
                //指定时间窗口 5s
                .timeWindow(Time.seconds(5))
                //对元组第二个属性求和
                .sum(1);

        dataStream.print();

        env.execute("Window WordCount");
    }

    public static class Splitter implements FlatMapFunction<String, Tuple2<String, Integer>> {
        private static final String DELIMITER = " ";
        @Override
        public void flatMap(String sentence, Collector<Tuple2<String, Integer>> out) throws Exception {
            for (String word: sentence.split(DELIMITER)) {
                out.collect(new Tuple2<>(word, 1));
            }
        }
    }
}
